modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/Walker.java

```diff
@@ -1115,22 +1106,9 @@ public ANode visitClassfuncref(ClassfuncrefContext ctx) {
 
     @Override
     public ANode visitConstructorfuncref(ConstructorfuncrefContext ctx) {
-        if (!ctx.decltype().LBRACE().isEmpty()) {
-            // array constructors are special: we need to make a synthetic method
-            // taking integer as argument and returning a new instance, and return a ref to that.
-            Location location = location(ctx);
-            String arrayType = ctx.decltype().getText();
-            SReturn code = new SReturn(location,
-                new ENewArray(location, arrayType, Arrays.asList(
-                    new EVariable(location, "size")), false));
-            String name = nextLambda();
-            globals.addSyntheticMethod(new SFunction(new FunctionReserved(), location, arrayType, name,
-                Arrays.asList("int"), Arrays.asList("size"), Arrays.asList(code), true));
-
-            return new EFunctionRef(location(ctx), "this", name);
-        }
-
-        return new EFunctionRef(location(ctx), ctx.decltype().getText(), ctx.NEW().getText());
+        return ctx.decltype().LBRACE().isEmpty() == false ?
+            new ENewArrayFunctionRef(location(ctx), ctx.decltype().getText()) :
```

Comment:
can we flip this to use positive logic? if empty, function ref, else array function ref

Reply:
Done.